src/thread.rs

```diff
@@ -147,6 +329,25 @@ impl<'mir, 'tcx> Thread<'mir, 'tcx> {
             b"<unnamed>"
         }
     }
+
+    /// Start the thread yielding. Returns true if this thread has watch metadata.
+    fn on_yield(&mut self, max_yield: u32) -> bool {
+        let iteration_count = self.yield_state.get_iteration_count();
+        let block = if max_yield == 0 {
+            // A value of 0 never blocks
+            false
+        } else {
+            iteration_count >= max_yield
+        };
+        if block {
+            log::trace!("Thread entered Blocking yield");
+            self.state = ThreadState::BlockedOnYield;
+        } else {
+            log::trace!("Thread entered standard yield with iteration {:?}", iteration_count);
+            self.state = ThreadState::DelayOnYield;
```

Comment:
So the purpose of this state seems to be to make the thread sleep until there is no other thread to make progress with any more, then it gets woken up again, and *then* we do `start_iteration` -- right?

What I do not understand is: this seems very heavy-handed for a yield. A yielded thread will *never be scheduled again* if the rest of the program just keeps making progress somewhere. That's very different from the old yield behavior. Couldn't this be a source of problems? Or am I misunderstanding what happens?